packages/internal/src/hex.js

```diff
@@ -0,0 +1,118 @@
+/**
+ * @typedef {object} HexCodec
+ * @property {(buf: Uint8Array) => string} encodeHex
+ * @property {(hex: string) => Uint8Array} decodeHex
+ */
+
+/**
+ * Create a hex codec that is portable across standard JS environments.
+ *
+ * @returns {HexCodec}
+ */
+export const makePortableHexCodec = () => {
+  /** @type {string[]} */
+  const encodings = Array.from({ length: 256 }, (_, b) =>
+    // Write the hex representation of the byte.
+    b.toString(16).padStart(2, '0'),
+  );
+
+  const uncurryThis = Function.bind.bind(Function.bind.call);
+
+  /**
+   * Define a set of mutations to apply to the hex digits. This is done to allow
+   * for case-insensitive hex decoding. The mutations are applied to the hex
+   * digits in a way that allows for all possible combinations of upper and
+   * lower case letters.
+   *
+   * @type {((c: string) => string)[]}
+   */
+  const mutations = [
+    uncurryThis(String.prototype.toLowerCase),
+    uncurryThis(String.prototype.toUpperCase),
+  ];
+
+  /**
+   * Create map entries for every permutation of case for each hex byte. The map
+   * is keyed by the hex string and the value is the byte value. This allows for
+   * fast lookups when decoding hex strings.
+   *
+   * @type {Map<string, number>}
+   */
+  const decodings = new Map(
+    encodings.flatMap((chars, b) =>
+      // Apply each mutation to the 0th character.
+      mutations.flatMap(f0 =>
+        // Apply each mutation to the 1st character.
+        mutations.map(
+          f1 => /** @type {const} */ ([f0(chars[0]) + f1(chars[1]), b]),
+        ),
+      ),
+    ),
+  );
```

Comment:
There's no need for initializing code to be so indirect.
```suggestion
  /**
   * Create map entries for every permutation of case for each hex byte. The map
   * is keyed by the hex string and the value is the byte value. This allows for
   * fast lookups when decoding hex strings.
   *
   * @type {Map<string, number>}
   */
  const decodings = new Map(
    encodings.flatMap((hh, b) => {
      const h1 = hh[0].toLowerCase();
      const h2 = hh[1].toLowerCase();
      const H1 = h1.toUpperCase();
      const H2 = h2.toUpperCase();
      return [
        [`${h1}${h2}`, b],
        [`${h1}${H2}`, b],
        [`${H1}${h2}`, b],
        [`${H1}${H2}`, b],
      ];
    }),
  );
```

It might also be better to promote the definitions of `encodings` and `decodings` outside of `makePortableHexCodec`, since they don't need specificity to any particular instantiation and are not expensive in terms of either CPU cycles or memory.

Reply:
Thanks for this less nested code.  I found the difference between `h[12]` and `H[12]` a little too subtle for me to easily distinguish while drowsy, but let me know what you think of my reinterpretation.

Reply:
Works for me!